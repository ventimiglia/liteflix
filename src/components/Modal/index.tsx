"use client";
import React, { ReactNode, useEffect, useState } from "react";
import Image from "next/image";
import Link from "next/link";
import { useRouter } from "next/navigation";
import { motion, AnimatePresence } from "framer-motion";

import Menu from "../Menu";
import { uploadMovie } from "@/services/movies/client";
import Wizard, { WizardStep } from "../Wizard";

const Modal = ({ children }: { children: ReactNode }) => {
  const router = useRouter();
  const [showModal, setShowModal] = useState(false);
  const [image, setImage] = useState<File | null>(null);
  const [title, setTitle] = useState("");
  const [uploadProgress, setUploadProgress] = useState(0);
  const [step, setStep] = useState(WizardStep.INITIAL);

  const [isFormValid, setIsFormValid] = useState(false);

  useEffect(() => {
    setIsFormValid(image !== null && title.trim() !== "");
  }, [image, title]);

  const handleImageChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const file = e.target.files && e.target.files[0];
    setStep(WizardStep.UPLOADINGFILE);

    const progressInterval = setInterval(() => {
      setUploadProgress((prevProgress) => prevProgress + 1);
    }, 20);

    setTimeout(() => {
      clearInterval(progressInterval);

      if (!file?.type.startsWith("image/")) {
        setStep(WizardStep.UPLOADFILEERROR);
      } else {
        setImage(file);
        setStep(WizardStep.UPLOADFILESUCCESS);
        setIsFormValid(image !== null && title.trim() !== "");
      }
      setUploadProgress(100);
    }, 2000);
  };

  const handleTitleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setTitle(e.target.value);
    setIsFormValid(image !== null && e.target.value.trim() !== "");
  };

  const handleSubmit = async (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();

    if (!!image && title.trim() !== "") {
      setStep(WizardStep.LOADING);
      const response = await uploadMovie({ image, title });
      if (!!response) {
        setStep(WizardStep.SUBMITED);
      } else {
        setStep(WizardStep.UPLOADFILEERROR);
      }
    }
  };

  const handleClose = () => {
    setShowModal(false);
    setImage(null);
    setTitle("");
    setUploadProgress(0);
    setStep(WizardStep.INITIAL);
    setIsFormValid(false);
    router.refresh();
  };

  const retry = () => {
    setStep(WizardStep.INITIAL);
    setUploadProgress(0);
    setIsFormValid(false);
    setImage(null);
  };

  return (
    <>
      <button
        className="flex items-center gap-3 transition-transform hover:scale-105 focus:scale-105"
        onClick={() => setShowModal(true)}
      >
        {children}
      </button>
      <AnimatePresence>
        {showModal && (
          <motion.div
            initial={{ opacity: 0 }}
            animate={{ opacity: 1 }}
            exit={{ opacity: 0 }}
            className="flex justify-center items-center overflow-x-hidden overflow-y-auto fixed inset-0 z-50 outline-none bg-gray-900/70 focus:outline-none"
          >
            <motion.div
              initial={{ zIndex: -1 }}
              animate={{ zIndex: 100 }}
              exit={{ zIndex: -1 }}
              className="relative w-full lg:min-w-[700px] mx-auto max-w-3xl h-full lg:h-fit"
            >
              <div className="p-6 lg:px-16 lg:py-12 shadow-lg relative flex flex-col w-full bg-secondary outline-none h-full lg:h-fit focus:outline-none">
                <Link href="/">
                  <button
                    className="hidden lg:flex absolute top-4 right-4 focus:outline-none transition-transform hover:scale-150"
                    onClick={handleClose}
                  >
                    <Image
                      src="/plus.svg"
                      alt="Cerrar"
                      width={24}
                      height={24}
                      className="rotate-45"
                    />
                  </button>
                </Link>
                <header className="flex justify-between lg:hidden">
                  <Menu>
                    <Image
                      src="/menu.svg"
                      alt="Menu"
                      width={26}
                      height={26}
                      className="scale-x-[-1] w-auto h-auto"
                    />
                  </Menu>
                  <Image
                    src="/logo.svg"
                    alt="Liteflix"
                    width={98}
                    height={28}
                    className="w-auto h-auto"
                  />
                  <button>
                    <Image
                      src="/avatar.svg"
                      alt="Profile"
                      width={36}
                      height={36}
                    />
                  </button>
                </header>
                <Wizard
                  step={step}
                  title={title}
                  handleImageChange={handleImageChange}
                  handleTitleChange={handleTitleChange}
                  handleSubmit={handleSubmit}
                  handleClose={handleClose}
                  retry={retry}
                  uploadProgress={uploadProgress}
                  isFormValid={isFormValid}
                />
              </div>
            </motion.div>
          </motion.div>
        )}
      </AnimatePresence>
    </>
  );
};

export default Modal;
