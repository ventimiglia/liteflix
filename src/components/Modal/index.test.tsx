import { render, fireEvent, waitFor } from "@testing-library/react";
import Modal from "./";

jest.mock("next/navigation", () => ({
  useRouter() {
    return {
      route: "/",
      refresh: jest.fn(),
    };
  },
}));

const children = "Open Modal";

describe("Modal", () => {
  it("renders modal button", () => {
    const { getByText } = render(<Modal>{children}</Modal>);
    const buttonElement = getByText(children);
    expect(buttonElement).toBeInTheDocument();
  });

  it("opens modal on button click", () => {
    const { getByText, container, getByRole } = render(
      <Modal>{children}</Modal>
    );
    const buttonElement = getByText(children);
    fireEvent.click(buttonElement);
    const modalElement = getByRole("modal");
    expect(modalElement).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it("closes modal on close button click", async () => {
    const { getByText, getByRole, getByAltText  } = render(<Modal>{children}</Modal>);
    const buttonElement = getByText(children);
    fireEvent.click(buttonElement);

    const modalElement = getByRole("modal");
    expect(modalElement).toBeInTheDocument();

    const closeModalButton = getByAltText("Cerrar");
    fireEvent.click(closeModalButton);

    await waitFor(() => {
      expect(modalElement).not.toBeInTheDocument();
    });
  });
});
