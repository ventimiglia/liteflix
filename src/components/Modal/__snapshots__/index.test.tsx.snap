// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal opens modal on button click 1`] = `
<div>
  <button
    class="flex items-center gap-3 transition-transform hover:scale-105 focus:scale-105"
  >
    Open Modal
  </button>
  <div
    class="flex justify-center items-center overflow-x-hidden overflow-y-auto fixed inset-0 z-50 outline-none bg-gray-900/70 focus:outline-none"
    role="modal"
    style="opacity: 0;"
  >
    <div
      class="relative w-full lg:min-w-[700px] mx-auto max-w-3xl h-full lg:h-fit"
      style="z-index: -1;"
    >
      <div
        class="p-6 lg:px-16 lg:py-12 shadow-lg relative flex flex-col w-full bg-secondary outline-none h-full lg:h-fit focus:outline-none"
      >
        <a
          href="/"
        >
          <button
            class="hidden lg:flex absolute top-4 right-4 focus:outline-none transition-transform hover:scale-150"
          >
            <img
              alt="Cerrar"
              class="rotate-45"
              data-nimg="1"
              decoding="async"
              height="24"
              loading="lazy"
              src="/plus.svg"
              style="color: transparent;"
              width="24"
            />
          </button>
        </a>
        <header
          class="flex justify-between lg:hidden"
        >
          <button
            class="flex items-center gap-3 transition-transform hover:scale-125 focus:scale-125"
          >
            <img
              alt="open menu"
              class="scale-x-[-1] w-auto h-auto"
              data-nimg="1"
              decoding="async"
              height="26"
              loading="lazy"
              src="/menu.svg"
              style="color: transparent;"
              width="26"
            />
          </button>
          <img
            alt="Liteflix"
            class="w-auto h-auto"
            data-nimg="1"
            decoding="async"
            height="28"
            loading="lazy"
            src="/logo.svg"
            style="color: transparent;"
            width="98"
          />
          <button>
            <img
              alt="Profile"
              data-nimg="1"
              decoding="async"
              height="36"
              loading="lazy"
              src="/avatar.svg"
              style="color: transparent;"
              width="36"
            />
          </button>
        </header>
        <h3
          class="text-2xl lg:text-xl text-center text-primary mb-[72px] mt-24 lg:mb-12 lg:mt-0"
        >
          AGREGAR PELICULA
        </h3>
        <form
          class="flex flex-col items-center lg:gap-12"
        >
          <label
            class="flex items-center relative justify-center gap-4 w-full h-20 lg:h-24 border-2 border-dashed cursor-pointer mb-14 lg:mb-0 transition-colors hover:bg-white/5"
            id="file"
          >
            <img
              alt="Agregar imagen"
              data-nimg="1"
              decoding="async"
              height="16"
              loading="lazy"
              src="/clip.svg"
              style="color: transparent;"
              width="16"
            />
            <p>
              <b>
                AGREGÁ UN ARCHIVO 
              </b>
              <span
                class="hidden lg:inline"
              >
                O ARRASTRALO Y SOLTALO AQUÍ
              </span>
            </p>
            <input
              accept="image/*"
              class="absolute w-full h-full cursor-pointer opacity-0"
              name="file"
              placeholder="Agrega un archivo"
              test-id="file-input"
              type="file"
            />
          </label>
          <input
            class="text-center bg-transparent border-b-2 w-60 mb-24 lg:mb-auto focus:outline-none"
            id="title"
            name="title"
            placeholder="Título"
          />
          <button
            class="w-60 mb-6 bg-white text-secondary text-lg py-4 opacity-30 cursor-not-allowed"
            disabled=""
            type="submit"
          >
            SUBIR PELICULA
          </button>
          <button
            class="lg:hidden text-lg w-60 border border-white/50 bg-transparent px-2 justify-center h-14 transition hover:scale-105 hover:bg-gray-800/75 active:bg-gray-900/75 focus:outline-none focus:ring focus:ring-gray-500"
          >
            SALIR
          </button>
        </form>
      </div>
    </div>
  </div>
</div>
`;
